rust/phoenix-sdk/src/coinbase_price_listener.rs

```diff
@@ -0,0 +1,148 @@
+use crate::{market_event_handler::SDKMarketEvent, orderbook::Orderbook};
+use coinbase_pro_rs::structs::reqs::OrderSide;
+use coinbase_pro_rs::{structs::wsfeed::*, WSFeed};
+use futures::StreamExt;
+use phoenix_types::enums::*;
+use rust_decimal::prelude::*;
+use std::{
+    collections::BTreeMap,
+    sync::{mpsc::Sender, Arc, RwLock},
+    thread,
+    thread::JoinHandle,
+};
+
+pub struct CoinbasePriceListener {
+    pub worker: JoinHandle<Option<()>>,
+}
+
+impl CoinbasePriceListener {
+    pub fn new(market_name: String, sender: Sender<Vec<SDKMarketEvent>>) -> Self {
+        let ladder = Arc::new(RwLock::new(Orderbook {
+            size_mult: 1.0,
+            price_mult: 1.0,
+            bids: BTreeMap::new(),
+            asks: BTreeMap::new(),
+        }));
+        let worker = thread::Builder::new()
+            .name("coinbase-ladder".to_string())
+            .spawn(move || Self::run(ladder, market_name, sender, false))
+            .unwrap();
+
+        Self { worker }
+    }
+
+    pub fn new_with_last_trade_price(
+        market_name: String,
+        sender: Sender<Vec<SDKMarketEvent>>,
+    ) -> Self {
+        let ladder = Arc::new(RwLock::new(Orderbook {
+            size_mult: 1.0,
+            price_mult: 1.0,
+            bids: BTreeMap::new(),
+            asks: BTreeMap::new(),
+        }));
+        let worker = thread::Builder::new()
+            .name("coinbase-ladder".to_string())
+            .spawn(move || Self::run(ladder, market_name, sender, true))
+            .unwrap();
+
+        Self { worker }
+    }
+
+    pub fn join(self) -> Option<()> {
+        self.worker.join().unwrap()
+    }
+
+    pub fn run(
+        ladder: Arc<RwLock<Orderbook<Decimal, f64>>>,
+        market_name: String,
+        sender: Sender<Vec<SDKMarketEvent>>,
+        use_ticker: bool,
+    ) -> Option<()> {
+        let rt = tokio::runtime::Runtime::new().unwrap();
+
+        println!("Connecting to Coinbase Websocket API");
+        let coinbase_ws_url = "wss://ws-feed.pro.coinbase.com";
+
+        let channel_type = if use_ticker {
+            ChannelType::Ticker
+        } else {
+            ChannelType::Level2
+        };
+        
+        let mut stream = rt
+            .block_on(WSFeed::connect(
+                coinbase_ws_url,
+                &[market_name.as_str()],
+                &[channel_type],
+            ))
+            .unwrap();
+
+        loop {
```

Comment:
with a loop like this, i think it will just silently die if connection gets closed. Any tips on making this more resilient?

Reply:
Couple things i would check:
1) is there a way to check that the stream is closed
2) you can wrap the whole function in another loop that performs the reconnect. if the it fails consecutively, set up an increasing timer for each iteration

In general this stuff isn't super obvious and there isn't a 1 size fits all solution

Reply:
got it, will merge for now and take a look tmrw